daemon/rename.go

```diff
@@ -0,0 +1,30 @@
+package daemon
+
+import (
+	"github.com/docker/docker/engine"
+)
+
+func (daemon *Daemon) ContainerRename(job *engine.Job) engine.Status {
+	if len(job.Args) != 2 {
+		return job.Errorf("usage: %s OLD_NAME NEW_NAME", job.Name)
+	}
+	old_name := job.Args[0]
+	new_name := job.Args[1]
+
+	container := daemon.Get(old_name)
+	if container == nil {
+		return job.Errorf("No such container: %s", old_name)
+	}
+
+	container.Lock()
+	if err := daemon.containerGraph.Delete(container.Name); err != nil {
+		return job.Errorf("Failed to delete container %q: %v", old_name, err)
+	}
+	if _, err := daemon.reserveName(container.ID, new_name); err != nil {
+		return job.Errorf("Error when allocating new name: %s", err)
```

Comment:
There is a major issue here, as **error is returned without releasing the lock** (same for the code path above):

```
$ docker run --name test busybox true
$ docker rename test " "
Error response from daemon: Error when allocating new name: Invalid container name ( ), only [a-zA-Z0-9][a-zA-Z0-9_.-] are allowed
FATA[0000] Error: failed to rename container named test
$ docker ps -a # Hangs (probably daemon in deadlock)
```

I think we should use `defer container.Unlock()` to play safe.